andela/ah-shakas
# #160577485 Users should be able to favourite and unfavourite articles

#### What does this PR do?
Have the implementation for favourating and unfavourating features.
#### Description of Task to be completed?

Have the following end points working 

```POST /api/articles/<slug>/favourite```
Which favourates an artcle by adding the user and article to a database table

```DELETE /api/articles/<slug>/favourite```
This implements unfavouratin by deleting the user and article pair from the table

Also when a user gets an article they get a field informing theme if they have favourited it or not.
#### How should this be manually tested

prerequisites:
- Have postgres installed and running on localhost.
- Have [pipenv](https://pipenv.readthedocs.io/en/latest/) installed.

Having met the above proceed as follows.

1. Clone this repo with the command:
  `git clone https://github.com/andela/ah-shakas.git`

 1. Cd into the ah-shakas directory:
`cd ah-shakas`

1. Checkout to the branch ft-pagination-160577489  with the command:
`git checkout ft-favourite-articles-160577485`

1. Create a database and export the database:
`createdb testdb`
 
1. Create a `.env` file and add the following in it:
`DB_NAME="testdb"`
1. Activate the pipenv:
`pipenv shell`

1. Install the dependencies:
`pipenv install`


1. Migrate the models to the db:
`./manage.py migrate`

1. Run the server:
`./manage.py runserver` 
From postman


1. Create a user by sending a post request to the "/api/user/" end point with the data body ` {"usernaane':"prefered name", "email":"validemail", "password":"password"}`.
The return data will include a token which we will require to create articles from the "/api/articles/" endpoint.

- To check that favourating works we need to create an article by sending a post request to the 
endpoint `/api/articles/`
with the paylod:
```
{
	"article":{
		"title":"some title1",
		"body":"this is the body for to test pagination"
	}
}
```
since it requires authorization on the headers you have to add
 *key*: Authorization *value*:`<token from above registration> `
*key*: Content-Type *value*: application/json 

send a post request to the `/api/articles/<slug>/favourite` endpoint with the slug you got from creating the article above. This endpoint does not require a payload body.
since it requires authorization on the headers you have to add
 *key*: Authorization *value*:`<token from above registration>` 
*key*: Content-Type *value*: application/json

 The returned payload should look similar to this:
```
{
    "article": {
        "title": "some title",
        "description": "",
        "body": "this is some body content",
        "slug": "some-title",
        "image_url": null,
        "author": {
            "username": "kevin",
            "bio": "Update your bio",
            "image_url": "image-url",
            "following": false,
            "created_at": "2018-10-17T10:24:52.491125+03:00",
            "updated_at": "2018-10-17T10:24:52.493490+03:00"
        },
        "created_at": "2018-10-17T10:28:22.461795+03:00",
        "updated_at": "2018-10-17T10:28:22.461819+03:00",
        "favourited": true
    },
    "message": "favourited"
}
``` 
- To test unfavourating Send DELETE request to "/api/articles/<slug>/favourite" endpoint with the same slug as above to test unfavourating.
The returned payload should look similar to this:
```
{
{
    "article": {
        "title": "some title",
        "description": "",
        "body": "this is some body content",
        "slug": "some-title",
        "image_url": null,
        "author": {
            "username": "kevin",
            "bio": "Update your bio",
            "image_url": "image-url",
            "following": false,
            "created_at": "2018-10-17T10:24:52.491125+03:00",
            "updated_at": "2018-10-17T10:24:52.493490+03:00"
        },
        "created_at": "2018-10-17T10:28:22.461795+03:00",
        "updated_at": "2018-10-17T10:28:22.461819+03:00",
        "favourited": false
    },
    "message": "unfavourited"
}
```


#### Any background context you want to provide?
This is an implemantation similar to bookmarking an article to access later.

#### What are the relevant pivotal tracker stories?
[#160577485](https://www.pivotaltracker.com/n/projects/2198533/stories/160577485)
#### Screenshots 

<img width="1074" alt="image" src="https://user-images.githubusercontent.com/30701246/47069892-b98e3780-d1f8-11e8-9f4d-4deaec1870e1.png">

<img width="1079" alt="image" src="https://user-images.githubusercontent.com/30701246/47070033-1689ed80-d1f9-11e8-9cc0-88f716b682d9.png">
